Src/WindowsServer/WindowsServer.Shared/Implementation/AzureRoleEnvironmentContextReader.cs

```diff
@@ -62,25 +86,76 @@ internal set
+            // and without interfering with any customer code which could be loading same/different version of Microsoft.WindowsAzure.ServiceRuntime.dll.
+            try
             {
-                RoleInstance roleInstance = roleEnvironment.CurrentRoleInstance;
-                if (roleInstance != null)
+                AppDomainSetup domaininfo = new AppDomainSetup();            
+                domaininfo.ApplicationBase = AzureRoleEnvironmentContextReader.BaseDirectory;                
+
+                // Create a new AppDomain
+                tempDomainToLoadAssembly = AppDomain.CreateDomain(tempDomainName, null, domaininfo);
+
+                // Load the RemoteWorker assembly to the new domain            
+                tempDomainToLoadAssembly.Load(typeof(AssemblyLoader).Assembly.FullName);
+
+                // Any method invoked on this object will be executed in the newly created AppDomain.
+                AssemblyLoader remoteWorker = (AssemblyLoader)tempDomainToLoadAssembly.CreateInstanceAndUnwrap(typeof(AssemblyLoader).Assembly.FullName, typeof(AssemblyLoader).FullName);
+                remoteWorker.AssemblyName = AzureRoleEnvironmentContextReader.AssemblyName;
+                remoteWorker.Culture = AzureRoleEnvironmentContextReader.Culture;
+                remoteWorker.PublicKeyToken = AzureRoleEnvironmentContextReader.PublicKeyToken;
+                remoteWorker.VersionsToAttempt = AzureRoleEnvironmentContextReader.VersionsToAttempt;
+                bool success = remoteWorker.ReadAndPopulateContextInformation(ref this.roleName, ref this.roleInstanceName);
+                if (success)
                 {
-                    this.roleInstanceName = roleEnvironment.CurrentRoleInstance.Id;
-                    Role role = roleInstance.Role;
-                    if (role != null)
+                    WindowsServerEventSource.Log.TroubleshootingMessageEvent("Successfully loaded assembly from remote worker in separate AppDomain");
+                }
+                else
+                {
+                    WindowsServerEventSource.Log.TroubleshootingMessageEvent("Failed loading assembly from remote worker in separate AppDomain. Application is assumed not be running in Azure Cloud service.");
+                }
+            }
+            catch (Exception ex)
+            {
+                WindowsServerEventSource.Log.TroubleshootingMessageEvent("AzureRoleEnvironmentContextReader Initialize failed : " + ex.ToString());
+            }
+            finally
+            {                
+                try
+                {
+                    if (tempDomainToLoadAssembly != null)
                     {
-                        this.roleName = role.Name;
-                    }
+                        AppDomain.Unload(tempDomainToLoadAssembly);
+                        long endTimeInTicks = Stopwatch.GetTimestamp();
```

Comment:
If you meant the time measurement, its not used for any functionality. Its just used in logging messages to see how much it took for the initialize to complete.